## 01. Importing libraries & ***ords_prods_merge*** dataframe

In [29]:
# Import libraries

import pandas as pd
import numpy as np
import os

#creating path
path=r'C:\Users\Olufemi\Downloads\CareerFoundry Data Analyics\Instacart Basket Analysis 03_02_2025'

#orders_products_combined dataframe
ords_prods_merge=pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge.pkl'))

## 02. Creating a subset

In [102]:
# Creating a subset due to the large size of the dataframe- first 1 million rows
df = ords_prods_merge[:1000000]

In [104]:
# Checking the shape of the subset
df.shape

(1000000, 17)

## 03. Defining a Function

In [33]:
def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [108]:
# Apply the function

df['price_range' ] = df.apply(price_label, axis=1)

C:\Users\Olufemi\AppData\Local\Temp\ipykernel_7276\779712503.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range' ] = df.apply(price_label, axis=1)


In [112]:
# Check the values of the new function
df['price_range']. value_counts(dropna = False)

price_range
Mid range product     673340
Low range product     314392
High range product     12268
Name: count, dtype: int64

In [114]:
# Check and return max value of prices
df['prices'].max()

25.0

## 04. IF-Statement with the loc() Function

In [119]:
# # create high range label
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [121]:
# create medium range label
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [123]:
# create low range label
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [125]:
# return value counts of price_range_loc
df['price_range_loc']. value_counts(dropna = False)

price_range_loc
Mid-range product     673340
Low-range product     314392
High-range product     12268
Name: count, dtype: int64

In [127]:
# create high range label on full ords_prods_merge
ords_prods_merge. loc[ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [129]:
# create medium range label on full ords_prods_merge
ords_prods_merge.loc[(ords_prods_merge['prices'] <= 15) & (ords_prods_merge['prices'] > 5), 'price_range_loc' ] = 'Mid-range product'

In [131]:
# create low range label on full ords_prods_merge
ords_prods_merge.loc[ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [133]:
# return value counts of price_range_loc
ords_prods_merge['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     21865987
Low-range product     10126321
High-range product      412551
Name: count, dtype: int64

In [135]:
# Which days do the most orders take place? 0=Saturday, 1=Sunday, 2=Monday, 3=Tuesday, 4=Wednesday, ...6=Friday
ords_prods_merge['orders_day_of_week']. value_counts (dropna = False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [137]:
# Creating a 'busiest day' column that shows busiest day, least busy, and regularly busy days.
result = []

for value in ords_prods_merge["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [139]:
# Checking the result
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy

In [141]:
#Combining the result to the dataframe to give it meaning. Creating the new summary column
ords_prods_merge['busiest_day' ] = result

In [143]:
# check value count of new summary column
ords_prods_merge['busiest_day']. value_counts(dropna = False)

busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

## 4.7 Task 
  
  ### Question 2
#### Suppose your clients have changed their minds about the labels you created in your “busiest_day” column. Now, they want “Busiest day” to become “Busiest days” (plural). This label should correspond with the two busiest days of the week as opposed to the single busiest day. At the same time, they’d also like to know the two slowest days. Create a new column for this using a suitable method.

In [148]:
# check which days are busiest and slowest

ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [146]:
# Creating the two busiest days and two slowest days column

result = []

for value in ords_prods_merge["orders_day_of_week"]:
  if value == 0 or value ==1:
    result.append("Busiest days")
  elif value == 3 or value== 4:
    result.append("Slowest days")
  else:
    result.append("Regularly busy")

In [150]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest day

In [152]:
# Creating the busiest_days column

ords_prods_merge['busiest_days'] = result

In [154]:
# Checking the results

ords_prods_merge['busiest_days'].value_counts(dropna = False)

busiest_days
Regularly busy    12916111
Busiest days      11864412
Slowest days       7624336
Name: count, dtype: int64

# Question 3:

## **Observations from the result my code generated:**
   
   ### **Three Categories of "Busiest Days" Exist in the Result:**

   #### "Regularly busy" has 12,916,111 occurrences.
   #### "Busiest days"   has 11,864,412 occurrences.
   #### "Slowest days"   has 7,624,336 occurrences

## **Distribution Insight:**
   #### The "Regularly busy" category has the highest count, which means most days fall into this category
   #### The "Busiest days" category is slightly lower but still significant.
   #### The "Slowest days" category has the least occurrences, indicating fewer days with minimal activity.

## **No Missing Values Observed.**

# **Question 4:**

## When too many users make Instacart orders at the same time, the app freezes. The senior technical officer at Instacart wants you to identify the busiest hours of the day. Rather than by hour, they want periods of time labeled “Most orders,” “Average orders,” and “Fewest orders.” Create a new column containing these labels called “busiest_period_of_day.”

In [162]:
#Checking the columns' names
print(ords_prods_merge.columns)

Index(['Unnamed: 0', 'order_id', 'user_id', 'order_number',
       'orders_day_of_week', 'order_hour_of_day', 'days_since_prior_order',
       'product_id', 'add_to_cart_order', 'reordered', 'product_name',
       'aisle_id', 'department_id', 'prices', 'merge_', 'price_range_loc',
       'busiest_day', 'busiest_days'],
      dtype='object')


In [164]:
# Getting the value counts for each hour of day:

ords_prods_merge['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

In [166]:
# Create an empty list and divide the hours of the day into 3 sections starting with the most orders

result = []

for value in ords_prods_merge["order_hour_of_day"]:
  if value == 10 or value == 11 or value == 12:
    result.append("Most orders")
  elif value == 17 or value == 14 or value == 15 or value == 13 or value == 16:
    result.append("Average orders")
  else:
    result.append("Fewest orders")

In [168]:
result

['Fewest orders',
 'Fewest orders',
 'Fewest orders',
 'Fewest orders',
 'Fewest orders',
 'Fewest orders',
 'Fewest orders',
 'Fewest orders',
 'Fewest orders',
 'Fewest orders',
 'Fewest orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Fewest orders',
 'Fewest orders',
 'Fewest orders',
 'Fewest orders',
 'Fewest orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Fewest orders',
 'Fewest orders',
 'Fewest orders',
 'Fewest orders',
 'Fewest orders',
 'Fewest orders',
 'Fewest orders',
 'Fewest orders',
 'Fewest orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Fewest orders',
 'Fewest orders',
 'Fewest orders',
 'Fewest orders',
 'Fewest orders',


In [170]:
# Naming the column

ords_prods_merge['busiest_period_of_day'] = result

In [174]:
# Printing the frequency for this new column

ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Average orders    12635090
Fewest orders     11653337
Most orders        8116432
Name: count, dtype: int64

In [176]:
# Exporting the dataframe in a pickle format

ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_updated.pkl'))